diff --git a/core/lib/Drupal/Core/Routing/UrlGenerator.php b/core/lib/Drupal/Core/Routing/UrlGenerator.php
index c0763fe0..edff46b1 100644
--- a/core/lib/Drupal/Core/Routing/UrlGenerator.php
+++ b/core/lib/Drupal/Core/Routing/UrlGenerator.php
@@ -206,7 +206,7 @@ class UrlGenerator implements UrlGeneratorInterface {
           // check requirement
           if (!preg_match('#^' . $token[2] . '$#', $mergedParams[$token[3]])) {
             $message = sprintf('Parameter "%s" for route "%s" must match "%s" ("%s" given) to generate a corresponding URL.', $token[3], $name, $token[2], $mergedParams[$token[3]]);
-            throw new InvalidParameterException($message);
+            //throw new InvalidParameterException($message);
           }
 
           $url = $token[1] . $mergedParams[$token[3]] . $url;
